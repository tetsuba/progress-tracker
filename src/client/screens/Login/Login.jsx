import React, { useState, useContext } from 'react'
import { useMutation } from '@apollo/react-hooks'
import { Col, Container, Row } from 'react-bootstrap'

// CONTEXT
import { AuthenticatedContext } from '../../context/AuthenticatedContext'

// MUTATION
import {
  LOGIN_MUTATION,
  SEND_PASSWORD_RESET_CONFIRMATION_MUTATION,
  VERIFY_EMAIL_MUTATION,
} from '../../api/user/user.mutation'

// COMPONENTS
import LoginForm from '../../components/Form/LoginForm'
import ForgotMyPasswordForm from '../../components/Form/ForgotMyPasswordForm'
import Loading from '../../components/Loading/Loading';

// UTILS
import { getLoginStatus, getLoginError } from '../../components/Form/form-utils'
import EmailVerificationForm from '../../components/Form/EmailVerificationForm';

export default function Login() {
  const [userLogin, userLoginOptions] = useMutation(LOGIN_MUTATION)
  const [verifyEmail, verifyEmailOptions] = useMutation(VERIFY_EMAIL_MUTATION)
  const [sendPasswordResetConfirmation, sendPasswordResetConfirmationOptions] = useMutation(
    SEND_PASSWORD_RESET_CONFIRMATION_MUTATION
  )
  const [hideLoginForm, setHideLoginForm] = useState(false)
  const { toggle: authenticateUser } = useContext(AuthenticatedContext)

  return (
    <Container className="pt-5">
      <Row>
        {
          {
            loading: <Loading />,
            login: (
              <LoginForm
                error={getLoginError(userLoginOptions.error)}
                hideLoginForm={() => setHideLoginForm(true)}
                handleSubmit={(options) => {
                  userLogin(options)
                    .then(({data}) => {
                      authenticateUser(data.userLogin.token)
                    })
                    .catch((err) => console.log('Login error'))
                }}
              />
            ),
            forgetMyPassword: (
              <ForgotMyPasswordForm
                showLoginForm={() => setHideLoginForm(false)}
                handleSubmit={(options) => {
                  sendPasswordResetConfirmation(options)
                    .catch(() => console.log('error'))
                }}
                error={sendPasswordResetConfirmationOptions.error}
              >
                <h3>Forgot my password</h3>
                <p>
                  Enter your email address and we will send you a link to reset
                  your password.
                </p>
              </ForgotMyPasswordForm>
            ),
            emailNotVerified: (
              <EmailVerificationForm
                error={verifyEmailOptions.error}
                handleSubmit={(options) => {
                  verifyEmail(options)
                    .catch(() => console.log('error'))
                }}
              />
            ),
            success: (
              <Col id="LoginSuccess">
                <h3>Please check your email.</h3>
              </Col>
            ),
          }[getLoginStatus(
            hideLoginForm,
            userLoginOptions,
            verifyEmailOptions,
            sendPasswordResetConfirmationOptions,
          )]
        }
      </Row>
    </Container>
  )
}
