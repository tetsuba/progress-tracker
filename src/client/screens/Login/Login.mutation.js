import gql from 'graphql-tag';

export const LOGIN_MUTATION = gql`
    mutation($input: LoginUserInput!) {
        userLogin(input: $input) {
            id
            firstName
            lastName
            token
            email
        }
    }
`;
