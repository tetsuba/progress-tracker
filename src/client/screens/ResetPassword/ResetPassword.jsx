import React from 'react';
import { Col, Container, Row } from 'react-bootstrap';
import { Link, useParams } from "react-router-dom";
import { useMutation, useQuery } from "@apollo/react-hooks";

// QUERIES
import { CONFIRM_TOKEN_QUERY } from "../../api/token/token.query";

// MUTATIONS
import { REST_PASSWORD_MUTATION } from "../../api/user/user.mutation";

// COMPONENTS
import Loading from "../../components/Loading/Loading";
import EmailPasswordResetForm from "../../components/Form/EmailPasswordResetForm";
import ResetPasswordForm from "../../components/Form/ResetPasswordForm";

// UTILS
import { getRestPasswordStatus } from "../../components/Form/form-utils";


const ResetPassword = () => {
    let { token } = useParams();
    token = decodeURIComponent(token);
    const variables = { token };
    const confirmation = useQuery(CONFIRM_TOKEN_QUERY, {variables});
    const [resetPassword, resetPasswordOptions] = useMutation(REST_PASSWORD_MUTATION);

    return (
        <Container>
            {{
                loading: <Loading />,
                form: (
                    <ResetPasswordForm
                        resetPassword={resetPassword}
                        token={token}
                    />
                ),
                success: (
                    <Row>
                        <h3>
                            Password updated. Please <Link to="/login">click here</Link> to login
                        </h3>
                    </Row>
                ),
                error: ( /* token expired */
                    <>
                        <Row className="mt-5">
                            <p>Your last request to reset password has expired. Please send another request to reset your password.</p>
                        </Row>
                        <Row className="mt-5">
                            <Col>
                                <EmailPasswordResetForm />
                            </Col>

                        </Row>
                    </>
                ),
            }[getRestPasswordStatus(confirmation, resetPasswordOptions)]}
        </Container>
    )
};

export default ResetPassword;
