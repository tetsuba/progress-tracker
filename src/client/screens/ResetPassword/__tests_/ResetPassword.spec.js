import { act } from 'react-dom/test-utils'
import { graphRenderer, delay, updateTextInput } from '../../../../test/testHelper'

// COMPONENTS
import ResetPassword from '../ResetPassword'

import {
  confirmTokenQueryError,
  confirmTokenQuerySuccess,
  passwordResetMockDataSuccess,
  resetPasswordSuccess,
} from './mockData';
import Login from '../../Login/Login';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ token: 'token1234' }),
  Link: ({ children }) => children,
}))

describe('<ResetPassword>', () => {
  describe('Initial render', () => {
    describe('Reset password token not expired', () => {
      it('should render reset password form', async () => {
        let wrapper

        await act(async () => {
          wrapper = graphRenderer(ResetPassword, [confirmTokenQuerySuccess], {})
          await delay()
        })

        wrapper.update()
        expect(wrapper.find(ResetPassword)).toMatchSnapshot()
      })
      describe('A user submits a new password', () => {
        it('should render "success"', async () => {
          let wrapper
          await act(async () => {
            wrapper = graphRenderer(ResetPassword, [confirmTokenQuerySuccess, resetPasswordSuccess], {})
            await delay()
          })

          wrapper.update()

          act(() => {
            updateTextInput(wrapper, 'newPassword', '1234')
          })

          wrapper.update()

          act(() => {
            updateTextInput(wrapper, 'confirmPassword', '1234')
          })

          wrapper.update()

          await act(async () => {
            await wrapper.find('#ResetPasswordForm').get(0).props.onSubmit({
              preventDefault: jest.fn(),
            })
          })

          wrapper.update()
          expect(wrapper.find(ResetPassword)).toMatchSnapshot()
        })
      })
    })

    describe('Reset password token expired', () => {
      let wrapper
      beforeEach(async () => {
        await act(async () => {
          wrapper = graphRenderer(ResetPassword, [confirmTokenQueryError, passwordResetMockDataSuccess], {})
          await delay()
        })

        wrapper.update()
      })

      it('should render forgot my password form', async () => {
        expect(wrapper.find(ResetPassword)).toMatchSnapshot()
      })

      describe('A user submits a valid email address', () => {
        it('should render "success"', async () => {
          act(() => {
            updateTextInput(wrapper, 'email', 'test@test.com')
          })

          wrapper.update()

          await act(async () => {
            await wrapper.find('#ForgotMyPasswordForm').get(0).props.onSubmit({
              preventDefault: jest.fn(),
            })
            await delay()
          })

          wrapper.update()
          expect(wrapper.find(ResetPassword)).toMatchSnapshot()
        })
      })
    })
  })
})
