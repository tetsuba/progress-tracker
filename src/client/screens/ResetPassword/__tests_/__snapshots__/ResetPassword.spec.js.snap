// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetPassword> Initial render Reset password token expired A user submits a valid email address should render "success" 1`] = `
<ResetPassword>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col>
            <div
              className="col"
            >
              <h3
                id="ResetPasswordSuccess"
              >
                Password updated. Please 
                <Link
                  to="/login"
                >
                  click here
                </Link>
                 
                to login
              </h3>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</ResetPassword>
`;

exports[`<ResetPassword> Initial render Reset password token expired should render forgot my password form 1`] = `
<ResetPassword>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <ForgotMyPasswordForm
        handleSubmit={[Function]}
      >
        <Box
          max={500}
        >
          <div
            style={
              Object {
                "backgroundColor": "#f8f9fa",
                "border": "solid 1px #ccc",
                "borderRadius": "8px",
                "margin": "0 auto",
                "maxWidth": 500,
                "padding": "20px 20px",
                "width": "100%",
              }
            }
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Col>
                  <div
                    className="col"
                  >
                    <h3
                      className="text-danger"
                    >
                      This session has expired!
                    </h3>
                    <p>
                      Enter your email address and we will send you a link to reset your password.
                    </p>
                    <Form
                      className="w-100"
                      id="ForgotMyPasswordForm"
                      inline={false}
                      onSubmit={[Function]}
                    >
                      <form
                        className="w-100"
                        id="ForgotMyPasswordForm"
                        onSubmit={[Function]}
                      >
                        <FormGroup
                          controlId="forgotMyPasswordEmail"
                        >
                          <div
                            className="form-group"
                          >
                            <FormControl
                              isInvalid={false}
                              name="email"
                              onChange={[Function]}
                              placeholder="@"
                              required={true}
                              type="email"
                              value=""
                            >
                              <input
                                className="form-control"
                                id="forgotMyPasswordEmail"
                                name="email"
                                onChange={[Function]}
                                placeholder="@"
                                required={true}
                                type="email"
                                value=""
                              />
                            </FormControl>
                            <Feedback
                              type="invalid"
                            >
                              <div
                                className="invalid-feedback"
                              />
                            </Feedback>
                            <FormText
                              className="text-muted"
                            >
                              <small
                                className="text-muted form-text"
                              >
                                We'll never share your email with anyone else.
                              </small>
                            </FormText>
                          </div>
                        </FormGroup>
                        <Button
                          active={false}
                          className="float-right"
                          disabled={false}
                          id="ForgotMyPasswordSubmit"
                          type="submit"
                          variant="primary"
                        >
                          <button
                            className="float-right btn btn-primary"
                            disabled={false}
                            id="ForgotMyPasswordSubmit"
                            type="submit"
                          >
                            Reset Password
                          </button>
                        </Button>
                      </form>
                    </Form>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </Box>
      </ForgotMyPasswordForm>
    </div>
  </Container>
</ResetPassword>
`;

exports[`<ResetPassword> Initial render Reset password token not expired A user submits a new password should render "success" 1`] = `
<ResetPassword>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <Row
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col>
            <div
              className="col"
            >
              <h3
                id="ResetPasswordSuccess"
              >
                Password updated. Please 
                <Link
                  to="/login"
                >
                  click here
                </Link>
                 
                to login
              </h3>
            </div>
          </Col>
        </div>
      </Row>
    </div>
  </Container>
</ResetPassword>
`;

exports[`<ResetPassword> Initial render Reset password token not expired should render reset password form 1`] = `
<ResetPassword>
  <Container
    className="pt-5"
    fluid={false}
  >
    <div
      className="pt-5 container"
    >
      <ResetPasswordForm
        handleSubmit={[Function]}
        token="token1234"
      >
        <Box
          max={500}
        >
          <div
            style={
              Object {
                "backgroundColor": "#f8f9fa",
                "border": "solid 1px #ccc",
                "borderRadius": "8px",
                "margin": "0 auto",
                "maxWidth": 500,
                "padding": "20px 20px",
                "width": "100%",
              }
            }
          >
            <Row
              noGutters={false}
            >
              <div
                className="row"
              >
                <Col>
                  <div
                    className="col"
                  >
                    <h3>
                      Reset Password
                    </h3>
                    <Form
                      className="w-100"
                      id="ResetPasswordForm"
                      inline={false}
                      onSubmit={[Function]}
                    >
                      <form
                        className="w-100"
                        id="ResetPasswordForm"
                        onSubmit={[Function]}
                      >
                        <FormGroup
                          className="mb-3"
                          controlId="newPassword"
                        >
                          <div
                            className="mb-3 form-group"
                          >
                            <FormLabel
                              className="text-muted"
                              column={false}
                              srOnly={false}
                            >
                              <label
                                className="text-muted form-label"
                                htmlFor="newPassword"
                              >
                                New password
                              </label>
                            </FormLabel>
                            <FormControl
                              name="newPassword"
                              onChange={[Function]}
                              placeholder="******"
                              required={true}
                              type="password"
                              value=""
                            >
                              <input
                                className="form-control"
                                id="newPassword"
                                name="newPassword"
                                onChange={[Function]}
                                placeholder="******"
                                required={true}
                                type="password"
                                value=""
                              />
                            </FormControl>
                            <PasswordStrength
                              password=""
                            >
                              <div>
                                <ProgressBar
                                  animated={false}
                                  isChild={false}
                                  label=""
                                  max={100}
                                  min={0}
                                  now={0}
                                  srOnly={false}
                                  striped={false}
                                  variant="danger"
                                >
                                  <div
                                    className="progress"
                                  >
                                    <div
                                      aria-valuemax={100}
                                      aria-valuemin={0}
                                      aria-valuenow={0}
                                      className="progress-bar bg-danger"
                                      role="progressbar"
                                      style={
                                        Object {
                                          "width": "0%",
                                        }
                                      }
                                    />
                                  </div>
                                </ProgressBar>
                                <small
                                  className="form-text text-muted"
                                  id="passwordHelpBlock"
                                />
                              </div>
                            </PasswordStrength>
                          </div>
                        </FormGroup>
                        <FormGroup
                          className="mb-3"
                          controlId="confirmPassword"
                        >
                          <div
                            className="mb-3 form-group"
                          >
                            <FormLabel
                              className="text-muted"
                              column={false}
                              srOnly={false}
                            >
                              <label
                                className="text-muted form-label"
                                htmlFor="confirmPassword"
                              >
                                Confirm password
                              </label>
                            </FormLabel>
                            <FormControl
                              isInvalid={false}
                              name="confirmPassword"
                              onChange={[Function]}
                              placeholder="******"
                              required={true}
                              type="password"
                              value=""
                            >
                              <input
                                className="form-control"
                                id="confirmPassword"
                                name="confirmPassword"
                                onChange={[Function]}
                                placeholder="******"
                                required={true}
                                type="password"
                                value=""
                              />
                            </FormControl>
                            <Feedback
                              type="invalid"
                            >
                              <div
                                className="invalid-feedback"
                              />
                            </Feedback>
                            <FormText
                              className="text-muted"
                            >
                              <small
                                className="text-muted form-text"
                              >
                                Matching password 0 of 0
                              </small>
                            </FormText>
                          </div>
                        </FormGroup>
                        <Button
                          active={false}
                          className="float-right"
                          disabled={false}
                          id="ResetPasswordSubmit"
                          type="submit"
                          variant="primary"
                        >
                          <button
                            className="float-right btn btn-primary"
                            disabled={false}
                            id="ResetPasswordSubmit"
                            type="submit"
                          >
                            Reset Password
                          </button>
                        </Button>
                      </form>
                    </Form>
                  </div>
                </Col>
              </div>
            </Row>
          </div>
        </Box>
      </ResetPasswordForm>
    </div>
  </Container>
</ResetPassword>
`;
